browser/ipfs/test/ipfs_service_browsertest.cc

```diff
@@ -1053,57 +1067,175 @@ IN_PROC_BROWSER_TEST_F(IpfsServiceBrowserTest, CannotLoadIPFSImageFromHTTP) {
+      browser()->tab_strip_model()->GetActiveWebContents();
+  EXPECT_EQ(
       contents->GetURL(),
-      net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES);
+      GetURL(
+          "b.com",
+          "/ipfs/bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq/"
+          "simple.html?a=b"));
+}
 
-  EXPECT_EQ(domain, gateway.host());
+IN_PROC_BROWSER_TEST_F(
+    IpfsServiceBrowserTest,
+    TopLevelAutoRedirectsOn_ASK_RedirectFromGatewayLikeUrl_IpfsPath) {
+  ResetTestServer(
+      base::BindRepeating(&IpfsServiceBrowserTest::HandlePublicGatewayRequest,
+                          base::Unretained(this)));
+  browser()->profile()->GetPrefs()->SetBoolean(kIPFSAutoRedirectGateway, true);
+  browser()->profile()->GetPrefs()->SetInteger(
+      kIPFSResolveMethod,
+      static_cast<int>(ipfs::IPFSResolveMethodTypes::IPFS_ASK));
+  GURL gateway = GetURL("b.com", "/");
+  SetIPFSDefaultGatewayForTest(gateway);
+
+  auto tab_url = GetURL(
+      "a.com",
+      "/ipfs/bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq/"
+      "simple.html?a=b");
+
+  ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), tab_url));
+  content::WebContents* contents =
+      browser()->tab_strip_model()->GetActiveWebContents();
+  EXPECT_EQ(
+      contents->GetURL(),
+      GURL("ipfs://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq/"
+           "simple.html?a=b"));
+}
+
+IN_PROC_BROWSER_TEST_F(
+    IpfsServiceBrowserTest,
+    TopLevelAutoRedirectsOn_ASK_RedirectFromGatewayLikeUrl_IpfsSubDomain) {
+  ResetTestServer(
+      base::BindRepeating(&IpfsServiceBrowserTest::HandlePublicGatewayRequest,
+                          base::Unretained(this)));
+  browser()->profile()->GetPrefs()->SetBoolean(kIPFSAutoRedirectGateway, true);
+  browser()->profile()->GetPrefs()->SetInteger(
+      kIPFSResolveMethod,
+      static_cast<int>(ipfs::IPFSResolveMethodTypes::IPFS_ASK));
+  GURL gateway = GetURL("b.com", "/");
+  SetIPFSDefaultGatewayForTest(gateway);
+
+  auto tab_url = GetURL(
+      "bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq.ipfs.a.com",
+      "/simple.html?a=b");
+
+  ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), tab_url));
+  content::WebContents* contents =
+      browser()->tab_strip_model()->GetActiveWebContents();
+  EXPECT_EQ(
+      contents->GetURL(),
+      GURL("ipfs://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq/"
+           "simple.html?a=b"));
 }
 
 IN_PROC_BROWSER_TEST_F(IpfsServiceBrowserTest,
-                       TopLevelAutoRedirectsOnWithQuery) {
+                       TopLevelAutoRedirectsOn_DoNotTranslateSimpleUrls) {
   ResetTestServer(
-      base::BindRepeating(&IpfsServiceBrowserTest::HandleEmbeddedSrvrRequest,
+      base::BindRepeating(&IpfsServiceBrowserTest::HandlePublicGatewayRequest,
                           base::Unretained(this)));
   browser()->profile()->GetPrefs()->SetBoolean(kIPFSAutoRedirectGateway, true);
```

Comment:
could you also add a test that if kIPFSAutoRedirectGateway is false, it does not translate the url?

Reply:
same if ipfs is disabled

Reply:
Fixed